onnx/onnxifi_wrapper.c

```diff
@@ -0,0 +1,870 @@
+  const onnxStatus status = event_wrapper->library->onnxReleaseEvent(
+    event_wrapper->event);
+
+  /*
+   * Note: ReleaseEvent either succeeded, or failed with internal error.
+   * Either way, it is not safe to use the event handle again.
+   */
+  memset(event_wrapper, 0, sizeof(struct onnxifi_event_wrapper));
+  free(event_wrapper);
+
+  return status;
+}
+
+
+ONNXIFI_PUBLIC onnxStatus ONNXIFI_ABI onnxInitGraph(
+    onnxBackend backend,
+    size_t onnxModelSize,
+    const void* onnxModel,
+    uint32_t weightsCount,
+    const onnxTensorDescriptor* weightDescriptors,
+    onnxGraph* graph)
+{
+  if (graph == NULL) {
+    return ONNXIFI_STATUS_INVALID_POINTER;
+  }
+  *graph = NULL;
+
+  if (backend == NULL) {
+    return ONNXIFI_STATUS_INVALID_BACKEND;
+  }
+
+  struct onnxifi_backend_wrapper* backend_wrapper =
+    (struct onnxifi_backend_wrapper*) backend;
+  if (backend_wrapper->magic != ONNXIFI_BACKEND_MAGIC) {
+    return ONNXIFI_STATUS_INVALID_BACKEND;
+  }
+  if (backend_wrapper->library == NULL) {
+    return ONNXIFI_STATUS_INVALID_BACKEND;
+  }
+
+  struct onnxifi_graph_wrapper* graph_wrapper =
+    (struct onnxifi_graph_wrapper*)
+      malloc(sizeof(struct onnxifi_graph_wrapper));
+  if (graph_wrapper == NULL) {
+    return ONNXIFI_STATUS_NO_SYSTEM_MEMORY;
+  }
+  memset(graph_wrapper, 0, sizeof(struct onnxifi_graph_wrapper));
+
+  /* Call onnxInitGraph with unwrapped backend handle */
+  const onnxStatus status = backend_wrapper->library->onnxInitGraph(
+    backend_wrapper->backend,
+    onnxModelSize,
+    onnxModel,
+    weightsCount,
+    weightDescriptors,
+    &graph_wrapper->graph);
+  switch (status) {
+    case ONNXIFI_STATUS_SUCCESS:
+    case ONNXIFI_STATUS_FALLBACK:
+      /* Success, return wrapped graph */
+      graph_wrapper->magic = ONNXIFI_GRAPH_MAGIC;
+      graph_wrapper->library = backend_wrapper->library;
+      *graph = (onnxGraph) graph_wrapper;
+      return status;
+    default:
+      /* Failure, release allocated memory */
+      free(graph_wrapper);
+      return status;
+  }
+}
+
+ONNXIFI_PUBLIC onnxStatus ONNXIFI_ABI onnxSetGraphIO(
+    onnxGraph graph,
+    uint32_t inputsCount,
+    const onnxTensorDescriptor* inputDescriptors,
+    uint32_t outputsCount,
+    const onnxTensorDescriptor* outputDescriptors)
+{
+  if (graph == NULL) {
+    return ONNXIFI_STATUS_INVALID_GRAPH;
+  }
+
+  struct onnxifi_graph_wrapper* graph_wrapper =
+    (struct onnxifi_graph_wrapper*) graph;
+  if (graph_wrapper->magic != ONNXIFI_GRAPH_MAGIC) {
+    return ONNXIFI_STATUS_INVALID_GRAPH;
+  }
+  if (graph_wrapper->library == NULL) {
+    return ONNXIFI_STATUS_INVALID_GRAPH;
+  }
+
+  /* Call onnxSetGraphIO with unwrapped graph handle */
+  return graph_wrapper->library->onnxSetGraphIO(
+    graph_wrapper->graph,
+    inputsCount,
+    inputDescriptors,
+    outputsCount,
+    outputDescriptors);
+}
+
+ONNXIFI_PUBLIC onnxStatus ONNXIFI_ABI
+  onnxRunGraph(
```

Comment:
What's the usage of this function?

Reply:
Typical use-case:
1. Create `graph` with `onnxInitGraph`
2. Set `inputFence.type = ONNXIFI_SYNCHRONIZATION_EVENT`
3. Create `inputFence.event` with `onnxInitEvent`
4. Set `outputFence.type = ONNXIFI_SYNCHRONIZATION_EVENT`
5. Set `outputFence.event = NULL` (this event will be created by `onnxRunGraph`)
6. Call `onnxRunGraph(graph, &inputFence, &outputFence)`